File: src/wasm/spatial_index.cpp
#include "spatial_index.h"
#include <cmath>
#include <algorithm>
#include <iostream>
#include <limits>
#include <vector>
#include <cstdint>
#include <cstring>

SpatialIndex::SpatialIndex() {}

SpatialIndex::~SpatialIndex() {
  cleanup();
}

void SpatialIndex::cleanup() {
  // Note: We don't delete the arrays here as they point to WASM memory
  // managed by the TypeScript side. We just reset pointers.
  cellOffsets = nullptr;
  cellItems = nullptr;
}

void SpatialIndex::initializeFromWasm(uint32_t cellOffsetsPtr, uint32_t cellItemsPtr, 
                   uint32_t cellOffsetsCount, uint32_t cellItemsCount,
                   int gridWidth, int gridHeight, float cellSize,
                   float minX, float minY, float maxX, float maxY,
                   void* wasmBuffer) {
  // Set up pointers to WASM memory
  this->cellOffsets = reinterpret_cast<uint32_t*>(cellOffsetsPtr);
  this->cellItems = reinterpret_cast<int32_t*>(cellItemsPtr);
  
  // Copy grid parameters
  this->cellOffsetsCount = cellOffsetsCount;
  this->cellItemsCount = cellItemsCount;
  this->gridWidth = gridWidth;
  this->gridHeight = gridHeight;
  this->cellSize = cellSize;
  this->minX = minX;
  this->minY = minY;
  this->maxX = maxX;
  this->maxY = maxY;
}

// Static helper: append items from cell without dedup
void SpatialIndex::addCellItemsNoDedup(const SpatialIndex* self, int cx, int cy, std::vector<int>& out) {
  if (!self) return;
  if (cx < 0 || cy < 0 || cx >= self->gridWidth || cy >= self->gridHeight) return;
  const int idx = cy * self->gridWidth + cx;
  if (idx < 0 || idx >= static_cast<int>(self->cellOffsetsCount) - 1) return;
  const uint32_t start = self->cellOffsets[idx];
  const uint32_t end = self->cellOffsets[idx + 1];
  for (uint32_t i = start; i < end; ++i) {
    out.push_back(self->cellItems[i]);
  }
}

// Static helper: append unique items from cell (linear check)
void SpatialIndex::addCellItemsUnique(const SpatialIndex* self, int cx, int cy, std::vector<int>& out) {
  if (!self) return;
  if (cx < 0 || cy < 0 || cx >= self->gridWidth || cy >= self->gridHeight) return;
  const int idx = cy * self->gridWidth + cx;
  if (idx < 0 || idx >= static_cast<int>(self->cellOffsetsCount) - 1) return;
  const uint32_t start = self->cellOffsets[idx];
  const uint32_t end = self->cellOffsets[idx + 1];
  for (uint32_t i = start; i < end; ++i) {
    const int item = self->cellItems[i];
    if (std::find(out.begin(), out.end(), item) == out.end()) {
      out.push_back(item);
    }
  }
}

RangeView SpatialIndex::query(Point2 p) const {
  if (cellOffsets == nullptr || cellItems == nullptr) {
    return RangeView();
  }

  const int cellX = static_cast<int>((p.x - minX) / cellSize);
  const int cellY = static_cast<int>((p.y - minY) / cellSize);

  if (cellX < 0 || cellX >= gridWidth || cellY < 0 || cellY >= gridHeight) {
    return RangeView();
  }

  // Compute boundary distances
  const float cellMinX = minX + cellX * cellSize;
  const float cellMaxX = cellMinX + cellSize;
  const float cellMinY = minY + cellY * cellSize;
  const float cellMaxY = cellMinY + cellSize;
  const float threshold = 0.1f;
  const bool nearLeft = (p.x - cellMinX) <= threshold;
  const bool nearRight = (cellMaxX - p.x) <= threshold;
  const bool nearBottom = (p.y - cellMinY) <= threshold;
  const bool nearTop = (cellMaxY - p.y) <= threshold;

  // Fast path: not near any boundary, return a zero-copy view onto the single cell
  if (!nearLeft && !nearRight && !nearBottom && !nearTop) {
    const int idx = cellY * gridWidth + cellX;
    if (idx >= 0 && idx < static_cast<int>(cellOffsetsCount) - 1) {
      const uint32_t start = cellOffsets[idx];
      const uint32_t end = cellOffsets[idx + 1];
      const uint32_t count = end - start;
      return RangeView(const_cast<int32_t*>(&cellItems[start]), count, false);
    }
    return RangeView();
  }

  // Slow path: union with adjacent cells, deduping and owning allocation
  std::vector<int> tmp;
  tmp.reserve(64); // likely small
  addCellItemsNoDedup(this, cellX, cellY, tmp);
  if (nearLeft)  addCellItemsUnique(this, cellX - 1, cellY, tmp);
  if (nearRight) addCellItemsUnique(this, cellX + 1, cellY, tmp);
  if (nearBottom) addCellItemsUnique(this, cellX, cellY - 1, tmp);
  if (nearTop)   addCellItemsUnique(this, cellX, cellY + 1, tmp);

  if (tmp.empty()) return RangeView();
  int32_t* buf = new int32_t[tmp.size()];
  std::memcpy(buf, tmp.data(), tmp.size() * sizeof(int32_t));
  return RangeView(buf, static_cast<uint32_t>(tmp.size()), true);
}

std::vector<int> SpatialIndex::queryArea(float areaMinX, float areaMinY, float areaMaxX, float areaMaxY) const {
  std::vector<int> results;
  
  if (cellOffsets == nullptr || cellItems == nullptr) {
    return results;
  }

  // Calculate cell bounds
  int startCellX = std::max(0, static_cast<int>((areaMinX - minX) / cellSize));
  int endCellX = std::min(gridWidth - 1, static_cast<int>((areaMaxX - minX) / cellSize));
  int startCellY = std::max(0, static_cast<int>((areaMinY - minY) / cellSize));
  int endCellY = std::min(gridHeight - 1, static_cast<int>((areaMaxY - minY) / cellSize));
  
  // Collect items from all cells in the area
  for (int cellY = startCellY; cellY <= endCellY; cellY++) {
    for (int cellX = startCellX; cellX <= endCellX; cellX++) {
      int cellIndex = cellY * gridWidth + cellX;
      
      if (cellIndex < 0 || cellIndex >= static_cast<int>(cellOffsetsCount) - 1) {
        continue;
      }

      uint32_t start = cellOffsets[cellIndex];
      uint32_t end = cellOffsets[cellIndex + 1];

      for (uint32_t i = start; i < end; ++i) {
        int itemId = cellItems[i];
        
        // Check for duplicates (can occur when items span multiple cells)
        if (std::find(results.begin(), results.end(), itemId) == results.end()) {
          results.push_back(itemId);
        }
      }
    }
  }

  return results;
}

File: src/logic/navmesh/SpatialIndex.ts
/**
 * Spatial index class that provides views into C++ spatial index data
 */
export class SpatialIndex {
  public cellOffsets: Uint32Array = new Uint32Array(0);
  public cellItems: Int32Array = new Int32Array(0);
  public gridWidth: number = 0;
  public gridHeight: number = 0;
  public cellSize: number = 0;
  public minX: number = 0;
  public minY: number = 0;
  public maxX: number = 0;
  public maxY: number = 0;

  constructor() {}

  public initializeFromWasm(
    cellOffsetsPtr: number,
    cellItemsPtr: number,
    cellOffsetsCount: number,
    cellItemsCount: number,
    gridWidth: number,
    gridHeight: number,
    cellSize: number,
    minX: number,
    minY: number,
    maxX: number,
    maxY: number,
    wasmMemory: ArrayBuffer
  ): void {
    // console.log(`[TS SPATIAL INIT DEBUG] Initializing spatial index:`);
    // console.log(`  - cellOffsetsPtr: ${cellOffsetsPtr}, cellItemsPtr: ${cellItemsPtr}`);
    // console.log(`  - cellOffsetsCount: ${cellOffsetsCount}, cellItemsCount: ${cellItemsCount}`);
    // console.log(`  - grid: ${gridWidth}x${gridHeight}, cellSize: ${cellSize}`);
    // console.log(`  - bbox: [${minX}, ${minY}, ${maxX}, ${maxY}]`);
    // console.log(`  - wasmMemory.byteLength: ${wasmMemory.byteLength}`);

    if (cellOffsetsPtr !== 0 && cellOffsetsCount > 0) {
      this.cellOffsets = new Uint32Array(wasmMemory, cellOffsetsPtr, cellOffsetsCount);
      // console.log(`[TS SPATIAL INIT DEBUG] cellOffsets created: length=${this.cellOffsets.length}, first few values: [${Array.from(this.cellOffsets.slice(0, 10)).join(', ')}]`);
    } else {
      // console.log(`[TS SPATIAL INIT DEBUG] cellOffsets not initialized (ptr=${cellOffsetsPtr}, count=${cellOffsetsCount})`);
    }

    if (cellItemsPtr !== 0 && cellItemsCount > 0) {
      this.cellItems = new Int32Array(wasmMemory, cellItemsPtr, cellItemsCount);
      // console.log(`[TS SPATIAL INIT DEBUG] cellItems created: length=${this.cellItems.length}`);
    } else {
      // console.log(`[TS SPATIAL INIT DEBUG] cellItems not initialized (ptr=${cellItemsPtr}, count=${cellItemsCount})`);
    }

    this.gridWidth = gridWidth;
    this.gridHeight = gridHeight;
    this.cellSize = cellSize;
    this.minX = minX;
    this.minY = minY;
    this.maxX = maxX;
    this.maxY = maxY;

    // console.log(`[TS SPATIAL INIT DEBUG] Initialization complete. Final state:`);
    // console.log(`  - cellOffsets.length: ${this.cellOffsets.length}, cellItems.length: ${this.cellItems.length}`);
    // console.log(`  - grid: ${this.gridWidth}x${this.gridHeight}, cellSize: ${this.cellSize}`);
    // console.log(`  - bbox: [${this.minX}, ${this.minY}, ${this.maxX}, ${this.maxY}]`);
  }

  public query(x: number, y: number): Int32Array {
    if (x < this.minX || x > this.maxX || y < this.minY || y > this.maxY) {
      // console.log(`[TS SPATIAL DEBUG] Out of bounds: x=${x} (${this.minX}-${this.maxX}), y=${y} (${this.minY}-${this.maxY})`);
      return new Int32Array(0);
    }

    const cx = Math.floor((x - this.minX) / this.cellSize);
    const cy = Math.floor((y - this.minY) / this.cellSize);

    const idx = cx + cy * this.gridWidth;
    if (idx < 0 || idx >= this.cellOffsets.length - 1) {
      return new Int32Array(0);
    }

    const threshold = 0.1;
    const cellMinX = this.minX + cx * this.cellSize;
    const cellMaxX = cellMinX + this.cellSize;
    const cellMinY = this.minY + cy * this.cellSize;
    const cellMaxY = cellMinY + this.cellSize;

    const nearLeft = (x - cellMinX) <= threshold;
    const nearRight = (cellMaxX - x) <= threshold;
    const nearBottom = (y - cellMinY) <= threshold;
    const nearTop = (cellMaxY - y) <= threshold;

    // Fast path: not near any boundary, return the primary cell slice directly
    if (!nearLeft && !nearRight && !nearBottom && !nearTop) {
      const start = this.cellOffsets[idx];
      const end = this.cellOffsets[idx + 1];
      return this.cellItems.subarray(start, end);
    }

    // Slow path: merge with adjacent cells, deduping
    const start = this.cellOffsets[idx];
    const end = this.cellOffsets[idx + 1];
    const base = this.cellItems.slice(start, end);
    const out: number[] = Array.from(base);
    const seen = new Set<number>(base as unknown as number[]);

    if (nearLeft) SpatialIndex.addCellItemsUnique(this, cx - 1, cy, seen, out);
    if (nearRight) SpatialIndex.addCellItemsUnique(this, cx + 1, cy, seen, out);
    if (nearBottom) SpatialIndex.addCellItemsUnique(this, cx, cy - 1, seen, out);
    if (nearTop) SpatialIndex.addCellItemsUnique(this, cx, cy + 1, seen, out);

    return Int32Array.from(out);
  }

  public getItemsInCell(cx: number, cy: number): Int32Array {
    const cellIndex = cx + cy * this.gridWidth;
    if (cellIndex < 0 || cellIndex >= this.cellOffsets.length - 1) {
      return new Int32Array(0);
    }

    const start = this.cellOffsets[cellIndex];
    const end = this.cellOffsets[cellIndex + 1];
    return this.cellItems.subarray(start, end);
  }

  private static addCellItemsUnique(self: SpatialIndex, cx: number, cy: number, seen: Set<number>, out: number[]): void {
    if (cx < 0 || cy < 0 || cx >= self.gridWidth || cy >= self.gridHeight) return;
    const idx = cx + cy * self.gridWidth;
    if (idx < 0 || idx >= self.cellOffsets.length - 1) return;
    const start = self.cellOffsets[idx];
    const end = self.cellOffsets[idx + 1];
    for (let i = start; i < end; i++) {
      const v = self.cellItems[i];
      if (!seen.has(v)) { seen.add(v); out.push(v); }
    }
  }
}
